packages/eslint-plugin/src/rules/strict-void-return.ts

```diff
@@ -0,0 +1,428 @@
+
+import { AST_NODE_TYPES } from '@typescript-eslint/utils';
+import * as tsutils from 'ts-api-utils';
+import * as ts from 'typescript';
+
+import * as util from '../util';
+
+type Options = [
+  {
+    allowReturnAny?: boolean;
+  },
+];
+
+type MessageId = `asyncFunc` | `nonVoidFunc` | `nonVoidReturn`;
+
+export default util.createRule<Options, MessageId>({
+  name: 'strict-void-return',
+  meta: {
+    type: 'problem',
+    docs: {
+      description:
+        'Disallow passing a value-returning function in a position accepting a void function',
+      requiresTypeChecking: true,
+    },
+    messages: {
+      asyncFunc:
+        'Async function used in a context where a void function is expected.',
+      nonVoidFunc:
+        'Value-returning function used in a context where a void function is expected.',
+      nonVoidReturn:
+        'Value returned in a context where a void return is expected.',
+    },
+    schema: [
+      {
+        type: 'object',
+        additionalProperties: false,
+        properties: {
+          allowReturnAny: {
+            type: 'boolean',
+            description:
+              'Whether to allow functions returning `any` to be used in place expecting a `void` function.',
+          },
+        },
+      },
+    ],
+  },
+  defaultOptions: [
+    {
+      allowReturnAny: false,
+    },
+  ],
+
+  create(context, [options]) {
+    const sourceCode = context.sourceCode;
+    const parserServices = util.getParserServices(context);
+    const checker = parserServices.program.getTypeChecker();
+
+    return {
+      ArrayExpression: (node): void => {
+        for (const elemNode of node.elements) {
+          if (
+            elemNode != null &&
+            elemNode.type !== AST_NODE_TYPES.SpreadElement
+          ) {
+            checkExpressionNode(elemNode);
+          }
+        }
+      },
+      ArrowFunctionExpression: (node): void => {
+        if (node.body.type !== AST_NODE_TYPES.BlockStatement) {
+          checkExpressionNode(node.body);
+        }
+      },
+      AssignmentExpression: (node): void => {
+        checkExpressionNode(node.right); // should ignore operators like `+=` or `-=` automatically
+      },
+      'CallExpression, NewExpression': checkFunctionCallNode,
+      JSXAttribute: (node): void => {
+        if (
+          node.value?.type === AST_NODE_TYPES.JSXExpressionContainer &&
+          node.value.expression.type !== AST_NODE_TYPES.JSXEmptyExpression
+        ) {
+          checkExpressionNode(node.value.expression);
+        }
+      },
+      MethodDefinition: checkClassMethodNode,
+      ObjectExpression: (node): void => {
+        for (const propNode of node.properties) {
+          if (propNode.type !== AST_NODE_TYPES.SpreadElement) {
+            checkObjectPropertyNode(propNode);
+          }
+        }
+      },
+      PropertyDefinition: checkClassPropertyNode,
+      ReturnStatement: (node): void => {
+        if (node.argument != null) {
```

Comment:
[Testing] Removing this `if` and just calling `checkExpressionNode(node.argument);` doesn't fail any unit tests. Unnecessary code, or missing tests?